src/mscorlib/src/System/Lazy.cs

```diff
@@ -50,63 +42,22 @@ internal static class LazyHelpers
     [ComVisible(false)]
     [DebuggerTypeProxy(typeof(System_LazyDebugView<>))]
     [DebuggerDisplay("ThreadSafetyMode={Mode}, IsValueCreated={IsValueCreated}, IsValueFaulted={IsValueFaulted}, Value={ValueForDebugDisplay}")]
-    public class Lazy<T>
+    public class Lazy<T> :
+        ISerializable
     {
-
-#region Inner classes
-        /// <summary>
-        /// wrapper class to box the initialized value, this is mainly created to avoid boxing/unboxing the value each time the value is called in case T is 
-        /// a value type
-        /// </summary>
-        [Serializable]
-        class Boxed
-        {
-            internal Boxed(T value)
-            {
-                m_value = value;
-            }
-            internal T m_value;
-        }
-
-
-        /// <summary>
-        /// Wrapper class to wrap the excpetion thrown by the value factory
-        /// </summary>
-        class LazyInternalExceptionHolder
+        private interface ILazyItem
         {
-            internal ExceptionDispatchInfo m_edi;
-            internal LazyInternalExceptionHolder(Exception ex)
-            {
-                m_edi = ExceptionDispatchInfo.Capture(ex);
-            }
+            T Value { get; }
+            bool IsValueCreated { get; }
+            bool IsValueFaulted { get; }
+            LazyThreadSafetyMode Mode { get; }
         }
-#endregion
-
-        // A dummy delegate used as a  :
-        // 1- Flag to avoid recursive call to Value in None and ExecutionAndPublication modes in m_valueFactory
-        // 2- Flag to m_threadSafeObj if ExecutionAndPublication mode and the value is known to be initialized
-        static readonly Func<T> ALREADY_INVOKED_SENTINEL = delegate 
-        {
-            Debug.Assert(false, "ALREADY_INVOKED_SENTINEL should never be invoked.");
-            return default(T);
-        };
-
-        //null --> value is not created
-        //m_value is Boxed --> the value is created, and m_value holds the value
-        //m_value is LazyExceptionHolder --> it holds an exception
-        private object m_boxed;
-
-        // The factory delegate that returns the value.
-        // In None and ExecutionAndPublication modes, this will be set to ALREADY_INVOKED_SENTINEL as a flag to avoid recursive calls
-        [NonSerialized]
-        private Func<T> m_valueFactory;
 
-        // null if it is not thread safe mode
-        // LazyHelpers.PUBLICATION_ONLY_SENTINEL if PublicationOnly mode
-        // object if ExecutionAndPublication mode (may be ALREADY_INVOKED_SENTINEL if the value is already initialized)
-        [NonSerialized]
-        private object m_threadSafeObj;
+        // m_implementation, a volatile reference, is set to null after m_value has been set
+        private volatile ILazyItem m_implementation;
```

Comment:
As the current implementation either holds a lock when updating `m_implementation` or we have no protection ==> `Lazy<'T>` can't be used concurrently then the only place where `volatile` matters is in the `.ctor` to ensure all writes to `ILazyItem` objects are visible to other threaads when they read it.